### ISSUE TITLE
[Question] Remove colons in the files located in the git repository

### ISSUE THREAD

**Issue by pm64** [searxng/searxng#2463]
Update / hints from the maintainers:

This discussion came from a MS-Windows issue and has been changed to a discussion about filenames and filesystem

- about "MS-Windows" please read this https://github.com/searxng/searxng/pull/5067#issuecomment-3130623406
- about "filenames and filesystem" see below https://github.com/searxng/searxng/issues/2463#issuecomment-1565338138

----

Hello, this report is not related to the features of SearXNG, but rather the structure of this repo.

It's not possible for Windows users to clone and checkout this repo due to the presence of a colon in a filename, which their OS doesn't support.

The file's path is `utils/templates/etc/httpd/sites-available/searxng.conf:socket`.

Here's the entire error message:

```
Cloning into 'C:\searxng'...
remote: Enumerating objects: 62036, done.        
remote: Counting objects: 100% (62036/62036), done.        
remote: Compressing objects: 100% (14226/14226), done.        
remote: Total 62036 (delta 38011), reused 61752 (delta 37944), pack-reused 0        
Receiving objects: 100% (62036/62036), 37.36 MiB | 19.60 MiB/s, done.
Resolving deltas: 100% (38011/38011), done.
error: invalid path 'utils/templates/etc/httpd/sites-available/searxng.conf:socket'
fatal: unable to checkout working tree
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'
```

**Comment by return42** [comment]
Sorry, but MS-Windows is not in the scope of this project.

**Comment by pm64** [comment]
Fair enough, let's forget Windows and focus only on best practices, independent of the OS. The use of colons in filenames can introduce a range of potential issues that make it a good practice to avoid them in a Git repository.

Unless the goal is to actively repel Windows users, who may just want to read and learn from (not build) this project, it might be worthwhile to consider this suggestion for a moment.

1. Portability Across Tools: Some software tools might interpret colons as command syntax or as part of a protocol (like http: or file:). This can cause unexpected behavior or errors when processing those files.

2. Readability and Simplicity: Filenames with colons can be harder to read and remember, particularly as the colon is often associated with specific meanings such as time notation or namespace separation in programming languages. This could potentially mislead developers about the file's content or purpose.

3. Ease of Use on the Command Line: In many command line interfaces, colons have specific meanings. For example, in Unix-like systems, the colon is used to separate items in the PATH variable. If a filename contains a colon, it may need to be escaped or quoted when used in command line operations, which can be cumbersome and error-prone.

4. Web Compatibility: As the SearXNG repo is hosted on a web-based platform, colons in filenames can cause problems. Browsers might interpret these colons as part of a protocol (like http:), leading to incorrect file paths in URLs. This can cause the file to not be displayed or downloaded correctly.

5. Avoiding Merge Conflicts: Different collaborators might have different conventions for using colons. If one collaborator creates a file with a colon in the name, and another creates a different file with the same name but without the colon, merging their changes could result in conflicts.

6. Potential Confusion with Time Stamps: A colon is often used to represent time (for example, 12:30), so using a colon in a filename could potentially lead to confusion if the filename is interpreted as a timestamp.

The offending file is located at utils/templates/etc/httpd/sites-available/searxng.conf:socket

**Comment by unixfox** [comment]
That's not even the colon the issue, is that the script installs a bunch of things that only work on Linux based systems. Even if you attempt to solve the colon issue, you will be still stuck with a failed install script.

Anyway you still have other solutions:
- Use WSL which runs linux inside your Windows: https://learn.microsoft.com/en-us/windows/wsl/install
- Follow one of these comments but for searxng (not recommended, use at your own risk!): 
	- https://github.com/searx/searx/issues/2085
	- https://github.com/searx/searx/discussions/2653

**Comment by pm64** [comment]
Not sure how to proceed. This issue should not be related to Windows, it's about best practices in Git. I provided a list of facts in support of this position, none of which have been addressed. And now the title has been changed to focus back on Windows, which again is completely unrelated to this ticket.

@unixfox, even if you insist on making this about Windows, you're still choosing to focus on building and running on Windows, without regard for users that might simply want to check out and study the code.

And, even if you only agree with 1 or 2 of the reasons that colons in filenames are considered bad practice, particularly in Web-based repositories, that should be reason enough to devote a moment of thoughtfulness to this trivial change.

Is it the position of this project's maintainers that the reasons I provided are not worthy of consideration? If so that's fine, but it would be helpful for myself and other users to actually say that. Otherwise it will appear that the substance of this issue has been ignored.

**Comment by unixfox** [comment]
@return42 If I'm not mistaken this [file](https://github.com/searxng/searxng/blob/master/utils/templates/etc/httpd/sites-available/searxng.conf%3Asocket) is copied only once at the installation with the script?

So if we rename it to a filename without a colon this would be fine? I think we shouldn't block Windows users from trying to use SearXNG with their own methods nor contributing to the project.

I've opened a PR for renaming the files with a colon: https://github.com/searxng/searxng/pull/2465

**Comment by return42** [comment]
> I've opened a PR for renaming the files with a colon: https://github.com/searxng/searxng/pull/2465

With my hints below we should check if this is the only file.

> Is it the position of this project's maintainers that the reasons I provided are not worthy of consideration? If so that's fine, but it would be helpful for myself and other users to actually say that. Otherwise it will appear that the substance of this issue has been ignored.

Please understand that sometimes topics are not immediately recognized in their depth. This discussion came from a Windows issue. But because we are typical Linux developers and the project is not supported on Windows, we ignored the problem.

With your second comment, you address a different topic, but the comment doesn't get to the heart of the problem either:

In a heterogeneous ecosystem there are [many different file systems](https://en.wikipedia.org/wiki/List_of_file_systems) and which file names and path names are supported always depends on the underlying kernel and the implementation of this file systems by the kernel.
To make it a little clearer:

- in a typical MS-Windows environment you will usually encounter e.g. NTFS file system, but there are also the old file systems like FAT / both do not support lower/upper case

- in a typical Linux environment (and its offshoots like android) you will usually find file systems like ext4 which support unicode but there are also old file systems like ext2 which do not support unicode

- old macOS used HFS+, HFS+ does not support unicode and comes in different flavors (most often lower/upper case is not supported) .. IDK know how its is with APFS

- ... the list is endless

So what is a good base that runs well on all systems.

I think ext4 is not a bad base as this file system is supported on all platforms / I have not used it on Windows yet, but as far as I know it can be accessed via WSL.

But here in this issue we see its not convinient for the Windows users.

What might be a better aternative for a common base? ..

I don't know myself, but I know that exFAT is a very widely used file system.
Maybe it makes sense to use the file naming rules of exFAT in the SearXNG project?  case insensitive and ... see https://en.wikipedia.org/wiki/ExFAT -->

![image](https://github.com/searxng/searxng/assets/554536/752b86c6-015c-4a65-a692-9d78fb1fdd9d)

**Comment by unixfox** [comment]
> > I've opened a PR for renaming the files with a colon: #2465
> 
> With my hints below we should check if this is the only file.

```
$ /tmp/searxng>$ find -type f -name "*:*"
./utils/templates/etc/uwsgi/apps-available/searxng.ini:socket
./utils/templates/etc/uwsgi/apps-archlinux/searxng.ini:socket
./utils/templates/etc/nginx/default.apps-available/searxng.conf:socket
./utils/templates/etc/httpd/sites-available/searxng.conf:socket
```

> > Is it the position of this project's maintainers that the reasons I provided are not worthy of consideration? If so that's fine, but it would be helpful for myself and other users to actually say that. Otherwise it will appear that the substance of this issue has been ignored.
> 
> Please understand that sometimes topics are not immediately recognized in their depth. This discussion came from a Windows issue. But because we are typical Linux developers and the project is not supported on Windows, we ignored the problem.

I'm for supporting the file systems for the main operating systems not the file systems for the "portable devices" like exFAT.

**Comment by return42** [comment]
> I'm for supporting the file systems for the main operating systems not the file systems for the "portable devices" like exFAT.
 
@unixfox not sure what mean by .. should https://github.com/searxng/searxng/pull/2465 be reviewed or closed?

**Comment by unixfox** [comment]
> > I'm for supporting the file systems for the main operating systems not the file systems for the "portable devices" like exFAT.
> 
> @unixfox not sure what mean by .. should #2465 be reviewed or closed?

If that doesn't affect the existing searxng installations with the script (you are technically the only one to know that), I'm for reviewing and merging that PR.

**Comment by LencoDigitexer** [comment]
Helped me

```bash
PS C:\github> git config --global core.protectNTFS false
PS C:\github> git clone https://github.com/searxng/searxng
Cloning into 'searxng'...
remote: Enumerating objects: 76003, done.
remote: Counting objects: 100% (7163/7163), done.
remote: Compressing objects: 100% (528/528), done.
remote: Total 76003 (delta 6658), reused 6671 (delta 6633), pack-reused 68840
Receiving objects: 100% (76003/76003), 56.62 MiB | 6.80 MiB/s, done.
Resolving deltas: 100% (44379/44379), done.
(base) PS C:\github>
```

**Comment by return42** [comment]
> [`git config --global core.protectNTFS false`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreprotectNTFS)

@LencoDigitexer thanks for pointing out :+1: .. I closed PR

- https://github.com/searxng/searxng/pull/2465#issuecomment-1985225033

I think with your hint we can close this issue.

**Comment by pm64** [comment]
Sorry @return42 but how does this resolve the issue?

From the docs on this git config option: "If set to true, do not allow checkout of paths that would cause problems with the NTFS filesystem"

So the solution is to potentially "cause problems with the NTFS filsystem"?

Why not just rename the offending files?

**Comment by return42** [comment]
> Why not just rename the offending files?

We do not have official support for MS-Win

> ", do not allow checkout " .. So the solution is to potentially "cause problems with the NTFS filsystem"?

and theses files are not needed on MS-Win.

**Comment by return42** [comment]
Since it is wanted I reopen this issue / please read my comment: https://github.com/searxng/searxng/pull/2465#issuecomment-1985537579

**Comment by KyleKolander** [comment]
This worked for me, so thanks @LencoDigitexer!    `git config --global core.protectNTFS false`

I was stumped on this for a while.  I didn't even pay attention to the clone output, or I'd have seen the error about the invalid path.  But in fairness, I've been doing this for a long time, and I've always at least been able to **_clone_** the repo to look through the code on my local.

I get what you're saying about not supporting Windows @return42 and @unixfox, but I'm 100% with @pm64 on this one.

It's not a matter of whether one can successfully execute scripts in the repo after cloning it, which it seems is where the hang up resides.  I don't believe anyone is asking you to support that.  It's just a matter of naming files in such a way that a user on a major operating system can clone the repo without errors.  I don't understand the resistance to renaming a file or two so that Windows users can simply _**clone**_ the repo and look through the code.

While the configuration provided by @LencoDigitexer works, I don't think closing the issue is the right move here.  It's not obvious and you'd be relying on people finding this issue, _**if**_ they even find it, after which they are already frustrated - not a good first experience.

Anyhow, do with this feedback what you will.  Regardless, cool project!

Thanks again @LencoDigitexer.  TIL

**Comment by BlockHackTk** [comment]
@KyleKolander
> This worked for me, so thanks @LencoDigitexer! git config --global core.protectNTFS false

Thank you for this workaround.

**Comment by notmydevice** [comment]
@KyleKolander, "git config --global core.protectNTFS false" may appear to work, but you're risking damage to your NTFS filesystem. So this is certainly not an acceptable workaround. The value of this issue is that it has revealed and documented a basic contempt for Windows users by this project's maintainers. Even if this specific issue is resolved, the narrow and exclusionary philosophy will continue to limit this project's potential, which is a pity.

**Comment by KyleKolander** [comment]
@BlockHackTk - you're welcome.

@notmydevice  - no disagreement here.  I think it was pretty clear from my post how I felt about it, but I also wanted to share how I was able to get it to work in case it helped someone else.  FWIW, I am not using it anymore, so I don't have a dog in this fight!  😄

**Comment by zc-devs** [comment]
My use case.

1. Clone the repo.
2. Make fixes for dockerfile and entrypoint.
3. Make a PR with changes.

> It's not a matter of whether one can successfully execute scripts in the repo after cloning it, which it seems is where the hang up resides. I don't believe anyone is asking you to support that. It's just a matter of naming files in such a way that a user on a major operating system can clone the repo without errors.

This! I do not use install script, nor want to run natively in Windows.

> limit this project's potential

After my changes for Kubernetes suited me well over a year, I wanted to offer them back to upstream.
I can't even clone the repo and almost two years (Karl!) have been passed in talks on [trivial PR](https://github.com/searxng/searxng/pull/2465) (is it a shame?). Good time to remember how/why Searx _NG_ was started ;)
Seems, my contributions are not welcomed here ¯\_(ツ)_/¯

https://github.com/searxng/searxng/pull/2465#issuecomment-1565452647
https://github.com/searxng/searxng/pull/2465#issuecomment-1985453732

(Microphone hit the floor, motivation left the room)

**Comment by return42** [comment]
> Seems, my contributions are not welcomed here ¯_(ツ)_/¯

Be assured, that is not the reason ..

> [trivial PR](https://github.com/searxng/searxng/pull/2465) (is it a shame?)

Yes PR is trivial, but its not the half of the work ..

> [This is just one point comes first into my mind .. may be there is more to do .. **and it needs a lot of testing** ..](https://github.com/searxng/searxng/pull/2465#issuecomment-1565406796)

In the long term we want to remove this part from SearXNG completely, until then it is not worth putting so much effort into it ... especially since there are enough workarounds for the "colon" issue.

**Comment by zc-devs** [comment]
I've subscribed to #2465 and can wait another couple of years, no big deal for me :)

**Comment by Vizonex** [comment]
@zc-devs You didn't even need to wait 5 more years seems I'm already doing it.

**Comment by zc-devs** [comment]
TLDR: ~Looking forward to your PR!~ Nice attempt!

<details>
  <summary>Offtopic</summary>

> I'm already doing it

While I need only the "clone" part, you went further. You are rock!

> You didn't even need to wait 5 more years

~Let's talk when PR is merged~ You see :)

> After my changes for Kubernetes suited me well over a year, I wanted to offer them back to upstream

There is nothing from me to offer now. Since around may, the image runs under unprivileged user by default and supports configuration via environment variables (e.g. SEARXNG_SECRET).

---

[OS: Cross-platform](https://en.wikipedia.org/wiki/Python_(programming_language))

[Support Tier 1](https://peps.python.org/pep-0011/#tier-1)

Yes, I'm looking at you, Ansible, as well.

</details>

**Comment by return42** [comment]
This discussion came from a MS-Windows issue and has been changed to a discussion about filenames and filesystem

- about "MS-Windows" please read this https://github.com/searxng/searxng/pull/5067#issuecomment-3130623406
- about "filenames and filesystem" see above https://github.com/searxng/searxng/issues/2463#issuecomment-1565338138

--- 

Everything has been said / I lock the conversation for now.
